File: services/queue_watcher.py
"""Наблюдатель очереди запуска Aternos (auto_confirm).

После server.start(), если сервер попал в очередь, фоновая задача каждые
15 секунд опрашивает статус через python-aternos (asyncio.to_thread) и, когда
сервер доходит до состояния «ожидает подтверждения», автоматически вызывает
server.confirm(). По ходу обновляет дашборды с позицией в очереди.

Таймаут — 15 минут; при неудаче подтверждения владелец получает уведомление.
"""

from __future__ import annotations

import asyncio
import logging
import time

from aiogram import Bot

import database
from services.aternos_api import AternosError, AternosManager
from services.dashboard import clear_queue_position, set_queue_position, update_dashboards

logger = logging.getLogger(__name__)

POLL_INTERVAL = 15      # сек
WATCH_TIMEOUT = 15 * 60  # 15 минут
MAX_CONSECUTIVE_FAILURES = 5  # подряд неудачных опросов — прекращаем наблюдение

# Активные наблюдатели: server_id -> корутина (предотвращает дубли).
_active_watchers: set[int] = set()


def is_watching(server_id: int) -> bool:
    """Идёт ли уже наблюдение за сервером (защита от повторного запуска)."""
    return server_id in _active_watchers


def start_queue_watcher(bot: Bot, owner_id: int, server_id: int) -> None:
    """Запускает фонового наблюдателя очереди (если ещё не запущен)."""
    if server_id in _active_watchers:
        return
    _active_watchers.add(server_id)
    asyncio.create_task(_watch_loop(bot, owner_id, server_id))


async def _watch_loop(bot: Bot, owner_id: int, server_id: int) -> None:
    """Основной цикл наблюдения; гарантирует очистку при любом исходе."""
    try:
        await _watch(bot, owner_id, server_id)
    except Exception:
        logger.exception("queue watcher для сервера %s упал", server_id)
    finally:
        _active_watchers.discard(server_id)


async def _watch(bot: Bot, owner_id: int, server_id: int) -> None:
    manager = AternosManager(owner_id)
    deadline = time.monotonic() + WATCH_TIMEOUT
    failures = 0
    last_dashboard_update = 0.0

    while time.monotonic() < deadline:
        try:
            status = await manager.get_queue_status(server_id)
            failures = 0
        except AternosError as exc:
            logger.warning("сервер %s: опрос очереди не удался: %s", server_id, exc)
            failures += 1
            if failures >= MAX_CONSECUTIVE_FAILURES:
                await _notify_owner(bot, owner_id, server_id, f"опрос очереди не удался: {exc}")
                break
            await asyncio.sleep(POLL_INTERVAL)
            continue

        raw_queue = status.get("queue")
        server_status = status.get("status")
        server_lang = str(status.get("lang") or "")
        logger.info(
            "server status: status=%r lang=%r queue raw=%r type=%s",
            server_status, server_lang, raw_queue, type(raw_queue).__name__,
        )
        queue_value = _parse_queue(raw_queue)

        # Aternos просит подтверждение запуска — ТОЛЬКО при явном сигнале.
        # Live-проверка (2026-08-13): lang='waiting'/'preparing' — обычная
        # очередь, сервер стартует сам; confirm там отвечает HTTP 400.
        if "confirm" in server_lang.lower():
            try:
                await manager.confirm_server(server_id)
                logger.info("сервер %s: запуск подтверждён (lang=%r)", server_id, server_lang)
                set_queue_position(server_id, 0)
                await _refresh_dashboard(bot, server_id)
            except AternosError as exc:
                logger.warning(
                    "сервер %s: confirm не удался (lang=%r): %s — ждём следующей итерации",
                    server_id, server_lang, exc,
                )
            await asyncio.sleep(POLL_INTERVAL)
            continue

        if server_lang in ("waiting", "preparing"):
            if queue_value is not None and queue_value > 0:
                set_queue_position(server_id, queue_value)
                logger.info("сервер %s: очередь идёт, позиция %s", server_id, queue_value)
            now = time.monotonic()
            if now - last_dashboard_update >= 30:
                await _refresh_dashboard(bot, server_id)
                last_dashboard_update = now
            await asyncio.sleep(POLL_INTERVAL)
            continue

        if server_lang in ("loading", "starting"):
            logger.info("сервер %s: грузится (lang=%r), ждём запуска", server_id, server_lang)
            await asyncio.sleep(POLL_INTERVAL)
            continue

        if server_lang in ("on", "online"):
            logger.info("сервер %s: сервер онлайн, watcher завершён", server_id)
            port = status.get("port")
            if port:
                await database.set_server_port(server_id, int(port))
                logger.info("сервер %s: сохранён mc-порт %s", server_id, port)
            clear_queue_position(server_id)
            await _refresh_dashboard(bot, server_id)
            break

        # Любое другое состояние — логируем и ждём следующую итерацию.
        logger.info("сервер %s: состояние lang=%r, ждём следующей итерации", server_id, server_lang)
        await asyncio.sleep(POLL_INTERVAL)
    else:
        clear_queue_position(server_id)
        await _notify_owner(bot, owner_id, server_id, "таймаут 15 минут: очередь не была пройдена")
        await _refresh_dashboard(bot, server_id)


def _parse_queue(raw_queue) -> int | None:
    """Позиция в очереди из lastStatus['queue'] (python-aternos 3.0.6).

    Значение может быть: int (в т.ч. -1 — ждёт подтверждения), числовая
    строка или dict вида {"count": N, "position": M}. Возвращает int либо
    None, когда очередь не определена (None / 0 / пусто / dict без чисел).
    """
    if raw_queue is None or raw_queue == "":
        return None
    if isinstance(raw_queue, bool):
        return None
    if isinstance(raw_queue, dict):
        for key in ("position", "count"):
            value = raw_queue.get(key)
            if isinstance(value, (int, float)) and not isinstance(value, bool):
                return int(value)
        logger.info("queue dict без числовой позиции: %r", raw_queue)
        return None
    try:
        return int(raw_queue)
    except (TypeError, ValueError):
        logger.warning(
            "неожиданное значение queue: %r (%s)", raw_queue, type(raw_queue).__name__
        )
        return None


async def _refresh_dashboard(bot: Bot, server_id: int) -> None:
    """Обновляет дашборды во всех чатах, где привязан сервер."""
    chats = await database.get_chats_for_server(server_id)
    if chats:
        await update_dashboards(bot)


async def _notify_owner(bot: Bot, owner_id: int, server_id: int, reason: str) -> None:
    """Уведомляет владельца о проблеме с подтверждением запуска."""
    server = await database.get_server(server_id)
    name = server["display_name"] if server else f"ID {server_id}"
    try:
        await bot.send_message(
            owner_id,
            f"⚠️ <b>Автоподтверждение запуска не сработало:</b>\n"
            f"🖥 {name}\n{reason}\n"
            "Проверьте Aternos вручную.",
        )
    except Exception:
        logger.warning("не удалось уведомить владельца %s о сервере %s", owner_id, server_id)
    await database.log_action(owner_id, "auto_confirm_failed", f"{name}: {reason}", server_id=server_id)

File: internal/queuewatcher/queuewatcher.go
// Package queuewatcher — background Aternos queue confirmer.
package queuewatcher

import (
	"context"
	"fmt"
	"html"
	"log"
	"strings"
	"sync"
	"time"

	tele "gopkg.in/telebot.v3"

	"mineops/internal/aternos"
	"mineops/internal/dashboard"
	"mineops/internal/database"
	"mineops/internal/i18n"
	"mineops/internal/util"
)

const (
	pollInterval         = 15 * time.Second
	idleTimeout          = 20 * time.Minute
	maxFailures          = 5
	dashboardPeriod      = 30 * time.Second
	blindConfirmInterval = 45 * time.Second
)

type Watcher struct {
	db       *database.DB
	managers *aternos.Registry
	dash     *dashboard.Dashboard

	mu     sync.Mutex
	active map[int64]bool
}

func New(db *database.DB, managers *aternos.Registry, dash *dashboard.Dashboard) *Watcher {
	return &Watcher{
		db:       db,
		managers: managers,
		dash:     dash,
		active:   make(map[int64]bool),
	}
}

func (w *Watcher) IsWatching(serverID int64) bool {
	w.mu.Lock()
	defer w.mu.Unlock()
	return w.active[serverID]
}

// Start launches the single background queue watcher (if not running yet):
// panel polling every 15 seconds, auto-confirm on lang=="confirm",
// plus a blind Confirm every 45 seconds until the server comes online.
func (w *Watcher) Start(b *tele.Bot, ownerID, serverID int64) {
	w.mu.Lock()
	if w.active[serverID] {
		w.mu.Unlock()
		return
	}
	w.active[serverID] = true
	w.mu.Unlock()
	go w.watchLoop(b, ownerID, serverID)
}

// Rescan starts watching all active servers that are already in the
// Aternos queue (survived a bot restart). Each server is checked against
// the panel once; if it is in the starting/queue state — a watcher
// takes over.
func (w *Watcher) Rescan(b *tele.Bot) {
	owners, err := w.db.GetAllOwners()
	if err != nil {
		log.Printf("queuewatcher: rescan: owners not fetched: %v", err)
		return
	}
	for _, owner := range owners {
		servers, err := w.db.GetActiveServersByOwner(owner.UserID)
		if err != nil {
			continue
		}
		manager := w.managers.For(owner.UserID)
		for _, s := range servers {
			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
			info, err := manager.FetchInfo(ctx, s.ID)
			cancel()
			if err != nil {
				continue
			}
			lang := strings.ToLower(info.Lang)
			if lang == "" || lang == "off" || lang == "offline" || lang == "stop" || lang == "stopped" {
				continue
			}
			w.Start(b, owner.UserID, s.ID)
		}
	}
}

func (w *Watcher) watchLoop(b *tele.Bot, ownerID, serverID int64) {
	defer func() {
		w.mu.Lock()
		delete(w.active, serverID)
		w.mu.Unlock()
		if r := recover(); r != nil {
			log.Printf("queue watcher %d: panic: %v", serverID, r)
		}
	}()

	manager := w.managers.For(ownerID)
	failures := 0
	lastDashUpdate := time.Time{}
	lastBlindConfirm := time.Time{}
	// inactiveSince — when the server was last in an "inactive" state
	// (not in queue, not loading). If it still has not started within
	// idleTimeout — the launch is considered failed and the owner is notified.
	var inactiveSince time.Time

	for {
		ctx, cancel := context.WithTimeout(context.Background(), 45*time.Second)
		status, err := manager.FetchInfo(ctx, serverID)
		cancel()

		if err != nil {
			log.Printf("queuewatcher: server %d: poll failed: %v", serverID, err)
			failures++
			if failures >= maxFailures {
				w.notifyOwner(b, ownerID, serverID, fmt.Sprintf("queue poll failed: %v", err))
				break
			}
			time.Sleep(pollInterval)
			continue
		}
		failures = 0

		serverLang := strings.ToLower(status.Lang)
		queueValue := parseQueue(status.Queue)

		// Server came online (status==1 or lang on/online) — notify chats
		// and finish watching.
		if status.Status == 1 || serverLang == "on" || serverLang == "online" {
			log.Printf("queuewatcher: server %d: online, watcher finished", serverID)
			if p := util.ToInt(status.Port); p > 0 {
				_ = w.db.SetServerPort(serverID, p)
			}
			w.dash.ClearQueuePosition(serverID)
			w.refreshDashboard(b, serverID)
			w.notifyServerOnline(b, serverID, status)
			return
		}

		// active — the server is in a state where launch is still in progress
		// (queue running, server loading, or awaiting confirmation).
		active := false

		// Aternos asks for launch confirmation — confirm right away.
		if strings.Contains(serverLang, "confirm") {
			active = true
			// Fresh context: the previous one was already canceled (cancel() after FetchInfo),
			// otherwise ConfirmServer fails immediately with context canceled.
			cctx, ccancel := context.WithTimeout(context.Background(), 45*time.Second)
			err := manager.ConfirmServer(cctx, serverID)
			ccancel()
			lastBlindConfirm = time.Now()
			if err != nil {
				log.Printf("queuewatcher: server %d: confirm failed (lang=%q): %v — waiting for next iteration",
					serverID, serverLang, err)
			} else {
				log.Printf("queuewatcher: server %d: start confirmed (lang=%q)", serverID, serverLang)
				w.dash.SetQueuePosition(serverID, 0)
				w.refreshDashboard(b, serverID)
			}
			time.Sleep(pollInterval)
			continue
		}

		// Blind Confirm every 45 seconds: Aternos does not always update
		// lang in time, and the confirm API is idempotent — an extra request does no harm.
		if time.Since(lastBlindConfirm) >= blindConfirmInterval {
			cctx, ccancel := context.WithTimeout(context.Background(), 45*time.Second)
			_ = manager.ConfirmServer(cctx, serverID)
			ccancel()
			lastBlindConfirm = time.Now()
			log.Printf("queuewatcher: server %d: blind confirm sent", serverID)
		}

		switch serverLang {
		case "waiting", "preparing":
			active = true
			if queueValue > 0 {
				w.dash.SetQueuePosition(serverID, queueValue)
				log.Printf("queuewatcher: server %d: queue in progress, position %d", serverID, queueValue)
			}
			now := time.Now()
			if now.Sub(lastDashUpdate) >= dashboardPeriod {
				w.refreshDashboard(b, serverID)
				lastDashUpdate = now
			}
		case "loading", "starting":
			active = true
			log.Printf("queuewatcher: server %d: loading (lang=%q), waiting for start", serverID, serverLang)
		default:
			log.Printf("queuewatcher: server %d: state lang=%q, waiting for next iteration", serverID, serverLang)
		}

		if active {
			inactiveSince = time.Time{}
		} else if inactiveSince.IsZero() {
			inactiveSince = time.Now()
		} else if time.Since(inactiveSince) >= idleTimeout {
			w.dash.ClearQueuePosition(serverID)
			w.notifyOwner(b, ownerID, serverID, "server did not start (20 min outside queue)")
			w.refreshDashboard(b, serverID)
			return
		}
		time.Sleep(pollInterval)
	}

	w.dash.ClearQueuePosition(serverID)
	w.refreshDashboard(b, serverID)
}

// parseQueue — queue position from lastStatus['queue'].
//
// The value can be: a number (including -1 — awaiting confirmation), a numeric
// string, or a dict like {"count": N, "position": M}. None — queue undefined.
func parseQueue(raw any) int {
	switch t := raw.(type) {
	case nil:
		return 0
	case string:
		if t == "" {
			return 0
		}
		return parseInt(t)
	case bool:
		return 0
	case float64:
		n := int(t)
		if n < 0 {
			return 0
		}
		return n
	case int:
		if t < 0 {
			return 0
		}
		return t
	case int64:
		n := int(t)
		if n < 0 {
			return 0
		}
		return n
	case map[string]any:
		for _, key := range []string{"position", "count"} {
			if v, ok := t[key]; ok {
				if n := parseIntFrom(v); n > 0 {
					return n
				}
			}
		}
		return 0
	}
	return 0
}

func parseInt(s string) int {
	n := 0
	fmt.Sscanf(s, "%d", &n)
	if n < 0 {
		return 0
	}
	return n
}

func parseIntFrom(v any) int {
	switch t := v.(type) {
	case float64:
		return int(t)
	case int:
		return t
	case int64:
		return int(t)
	case string:
		return parseInt(t)
	}
	return 0
}

func (w *Watcher) refreshDashboard(b *tele.Bot, serverID int64) {
	chats, err := w.db.GetChatsForServer(serverID)
	if err != nil {
		return
	}
	if len(chats) > 0 {
		w.dash.UpdateDashboards(context.Background(), b)
	}
}

func (w *Watcher) ownerLang(ownerID int64) string {
	if o, _ := w.db.GetOwner(ownerID); o != nil && o.Lang != "" {
		return o.Lang
	}
	return "ru"
}

func (w *Watcher) chatLang(chatID int64) string {
	if o, _ := w.db.GetChatOwner(chatID); o != nil && o.Lang != "" {
		return o.Lang
	}
	return "ru"
}

// Temporary notification, deleted after 60 seconds.
func (w *Watcher) notifyServerOnline(b *tele.Bot, serverID int64, status *aternos.ServerInfo) {
	server, err := w.db.GetServer(serverID)
	if err != nil || server == nil {
		return
	}
	name := server.DisplayName
	if name == "" {
		name = fmt.Sprintf("ID %d", serverID)
	}
	ip := server.ServerIP
	if ip == "" {
		ip = util.ToStr(status.IP)
	}
	if ip == "" {
		ip = i18n.T("ru", "not_set")
	}
	chats, err := w.db.GetChatsForServer(serverID)
	if err != nil {
		return
	}
	for _, ch := range chats {
		lang := w.chatLang(ch.ChatID)
		text := i18n.T(lang, "qw_online", html.EscapeString(name), html.EscapeString(ip))
		msg, err := b.Send(&tele.Chat{ID: ch.ChatID}, text)
		if err != nil {
			log.Printf("queuewatcher: online notification for server %d not sent to chat %d: %v",
				serverID, ch.ChatID, err)
			continue
		}
		go func(m *tele.Message) {
			time.Sleep(60 * time.Second)
			_ = b.Delete(m)
		}(msg)
	}
}

func (w *Watcher) notifyOwner(b *tele.Bot, ownerID, serverID int64, reason string) {
	server, err := w.db.GetServer(serverID)
	name := fmt.Sprintf("ID %d", serverID)
	if err == nil && server != nil {
		name = server.DisplayName
	}
	lang := w.ownerLang(ownerID)
	_, err = b.Send(&tele.Chat{ID: ownerID},
		i18n.T(lang, "qw_failed", html.EscapeString(name), reason))
	if err != nil {
		log.Printf("queuewatcher: failed to notify owner %d about server %d: %v", ownerID, serverID, err)
	}
	_ = w.db.LogAction(ownerID, "auto_confirm_failed", fmt.Sprintf("%s: %s", name, reason), 0, 0, serverID)
}
